libsplinter/src/service/network/interconnect/mod.rs

```diff
@@ -13,3 +13,707 @@
+
+        Ok(ServiceInterconnect {
+            dispatched_sender: dispatched_sender.clone(),
+            recv_join_handle,
+            send_join_handle,
+            shutdown_handle: ShutdownHandle {
+                sender: dispatched_sender,
+                dispatch_shutdown: service_msg_dispatcher_shutdown,
+            },
+        })
+    }
+}
+
+fn run_recv_loop<R>(
+    service_connector: &dyn ServiceLookup,
+    message_receiver: R,
+    dispatch_msg_sender: DispatchMessageSender<ComponentMessageType, ConnectionId>,
+) -> Result<(), String>
+where
+    R: MatrixReceiver + 'static,
+{
+    let mut connection_id_to_service_id: HashMap<String, String> = HashMap::new();
+    loop {
+        // receive messages from components
+        let envelope = match message_receiver.recv() {
+            Ok(envelope) => envelope,
+            Err(MatrixRecvError::Shutdown) => {
+                info!("Matrix has shutdown");
+                break Ok(());
+            }
+            Err(MatrixRecvError::Disconnected) => {
+                break Err("Unable to receive message: disconnected".into());
+            }
+            Err(MatrixRecvError::InternalError { context, .. }) => {
+                break Err(format!("Unable to receive message: {}", context));
+            }
+        };
+
+        let connection_id = envelope.id();
+        let service_id = if let Some(service_id) = connection_id_to_service_id.get(connection_id) {
+            Some(service_id.to_owned())
+        } else if let Some(service_id) = service_connector
+            .service_id(connection_id)
+            .map_err(|err| format!("Unable to get service id for {}: {}", connection_id, err))?
+        {
+            connection_id_to_service_id.insert(connection_id.to_string(), service_id.clone());
+            Some(service_id)
+        } else {
+            None
+        };
+
+        // If we have the service, pass message to dispatcher, else print error
+        if let Some(service_id) = service_id {
+            let mut component_msg: ComponentMessage =
+                match protobuf::parse_from_bytes(&envelope.payload()) {
+                    Ok(msg) => msg,
+                    Err(err) => {
+                        error!("Unable to dispatch message: {}", err);
+                        continue;
+                    }
+                };
+
+            trace!(
+                "Received message from {}: {:?}",
+                service_id,
+                component_msg.get_message_type()
+            );
+            match dispatch_msg_sender.send(
+                component_msg.get_message_type(),
+                component_msg.take_payload(),
+                service_id.into(),
+            ) {
+                Ok(()) => (),
+                Err((message_type, _, _)) => {
+                    error!("Unable to dispatch message of type {:?}", message_type)
+                }
+            }
+        } else {
+            error!("Received message from unknown service");
```

Comment:
Is this an error? Seems like something that could just be ignored, so maybe it should be logged at a lower level.

Reply:
Actually, it probably is an error - it means that the system is receiving messages from a services that either didn't register, authorize, or we decided we no longer needed. It may indicate that a resource was properly cleaned up or messages were not guarded before reaching this process.